File: Data structures/Problems/leetcode/0020_validParenthesis.js
var q01 = "()";
var q02 = "()[]{}";
var q03 = "({[]})";
var q04 = "[()]{}";
var q05 = "{[}]";
var q06 = "({[})";
var q07 = "(]";
var q08 = "[";
var q09 = "]";
var q10 = "";
var q11 = "((";
var q12 = "))";
var q13 = "([]))";
var q14 = "{(})";
var q15 = "([)]";
var q16 = "{([])}";
var q17 = "{{{{}}";
var q18 = "(((({{))";
var q19 = "{[()]";
var q20 = "(())[]{}";
var q_list = [q01, q02, q03, q04, q05, q06, q07, q08, q09, q10, q11, q12, q13, q14, q15, q16, q17, q18, q19, q20];
var isValid = function (s) {
    var stack = [];
    var dict = {
        "}": "{",
        ")": "(",
        "]": "["
    };
    for (var _i = 0, s_1 = s; _i < s_1.length; _i++) {
        var letter = s_1[_i];
        if (letter in dict) {
            if (stack.length > 0 && dict[letter] === stack[stack.length - 1]) {
                stack.pop();
            }
            else {
                return false;
            }
        }
        else {
            stack.push(letter);
        }
    }
    return stack.length > 0 ? false : true;
};
for (var _i = 0, q_list_1 = q_list; _i < q_list_1.length; _i++) {
    var question = q_list_1[_i];
    console.log(question);
    console.log(isValid(question));
}
// function isValid(s:string):boolean{
//     return true
// }

File: data_structures_and_algorithms/Problems/leetcode/0020_validParenthesis.py
q01 = "()"
q02 = "()[]{}"
q03 = "({[]})"
q04 = "[()]{}"
q05 = "{[}]"
q06 = "({[})"
q07 = "(]"
q08 = "["
q09 = "]"
q10 = ""
q11 = "(("
q12 = "))"
q13 = "([]))"
q14 = "{(})"
q15 = "([)]"
q16 = "{([])}"
q17 = "{{{{}}"
q18 = "(((({{))"
q19 = "{[()]"
q20 = "(())[]{}"

q_list = [q01,q02,q03,q04,q05,q06,q07,q08,q09,q10,q11,q12,q13,q14,q15,q16,q17,q18,q19,q20]


class Solution:
    def isValid(self, s: str) -> bool:
        # edge case:
        if len(s) % 2 == 1:
            return False   

        if s == "":
            return True   

        stack = []
        dict = {
            "}":"{",
            ")":"(",
            "]":"["
        }

        for letter in s:
            if letter in dict:
                if stack and stack[-1] == dict[letter]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(letter)
        return True if not stack else False




sol = Solution()

q = p = 1
for question in q_list:
    print(f"Q{q}) " +  f"{sol.isValid(question)}" + f" {q_list[q-1]}")
    q += 1

class Solution2():
    def isValid(self, s: str) -> bool:
        #edge case
        
        if len(s) % 2 == 1:
            return False
        if s == "":
            return True
        
        o_bracket = ("(", "[", "{") #use the index of the tuple to check 
        c_bracket = (")", "]", "}")
        stack = []

        for letter in s:
            if letter in c_bracket:
                if stack and o_bracket.index(stack[-1]) == c_bracket.index(letter):
                    stack.pop()
                else:
                    return False
            else:
                stack.append(letter)
        return True if not stack else False


sol2 = Solution2()

print(" - - - - ")
for question in q_list:
    print(f"Q{p}) " +  f"{sol2.isValid(question)}" + f" {q_list[p-1]}")
    p += 1


# q01: "()",        Output: True
# q02: "()[]{}",    Output: True
# q03: "({[]})",    Output: True
# q04: "[()]{}",    Output: True
# q05: "{[}]",      Output: False
# q06: "({[})",     Output: False
# q07: "(]",        Output: False
# q08: "[",         Output: False
# q09: "]",         Output: False
# q10: "",          Output: True
# q11: "((",        Output: False
# q12: "))",        Output: False
# q13: "([]))",     Output: False
# q14: "{(})",      Output: False
# q15: "([)]",      Output: False
# q16: "{([])}",    Output: True
# q17: "{{{{}}",    Output: False
# q18: "(((({{))",  Output: False
# q19: "{[()]",     Output: False
# q20: "(())[]{}",  Output: True